network/src/protocols/rpc/mod.rs

```diff
@@ -169,85 +182,197 @@ pub struct Rpc<TSubstream> {
+                        self.handle_outbound_rpc(req, &mut outbound_rpc_tasks).await;
+                    } else {
+                        break;
+                    }
+                },
+                _ = inbound_rpc_tasks.select_next_some() => {
+                },
+                request_id = outbound_rpc_tasks.select_next_some() => {
+                    // Remove request_id from pending_outbound_rpcs if not already removed.
+                    let _ = self.pending_outbound_rpcs.remove(&request_id);
+                }
+            }
+        }
+        info!(
+            "Rpc actor terminated for peer: {}",
+            self.peer_handle.peer_id().short_str()
+        );
     }
-}
 
-/// Handle an outbound rpc request event. Open a new substream then run the
-/// outbound rpc protocol over the substream.
-///
-/// The request results (including errors) are propagated up to the rpc client
-/// through the [`req.res_tx`] oneshot channel. Cancellation is done by the client
-/// dropping the receiver side of the [`req.res_tx`] oneshot channel. If the
-/// request is canceled, the substream will be dropped and a RST frame will be
-/// sent over the muxer closing the substream.
-///
-/// [`req.res_tx`]: OutboundRpcRequest::res_tx
-async fn handle_outbound_rpc<TSubstream>(peer_handle: PeerHandle<TSubstream>, req: RpcRequest)
-where
-    TSubstream: AsyncRead + AsyncWrite + Send + Unpin,
-{
-    match req {
-        RpcRequest::SendRpc(req) => {
-            let protocol = req.protocol;
-            let req_data = req.data;
-            let mut res_tx = req.res_tx;
-            let timeout = req.timeout;
-            let peer_id_str = peer_handle.peer_id().short_str();
-            // Future to run the actual outbound rpc protocol and get the results.
+    /// Handle a new inbound message. If the message is a response to a pending outbound RPC, send
+    /// response to upstream client and return immediately. Otherwise, it is an inbound rpc
+    /// request. In that case, we spawn the request in a new task and return.
+    fn handle_inbound_message(
+        &mut self,
+        notif: PeerNotification,
+        inbound_rpc_tasks: &mut FuturesUnordered<BoxFuture<'static, ()>>,
+    ) {
+        let notification_tx = self.rpc_handler_tx.clone();
+        let peer_handle = self.peer_handle.clone();
+        let peer_id_str = peer_handle.peer_id().short_str();
+        match notif {
+            PeerNotification::NewMessage(message) => {
+                match message {
+                    // This is a response to a pending outbound RPC.
+                    NetworkMessage::RpcResponse(response) => {
+                        if let Some(response_tx) =
+                            self.pending_outbound_rpcs.remove(&response.request_id)
+                        {
+                            if let Err(e) = response_tx.send(response) {
+                                // TODO: add ability to log protocol id as well?
```

Comment:
Good idea, it isn't entirely clear which message failed here. Could we also encode the request id? Can we send the request id back to the sender?

Reply:
We do send request Id back to the sender. Maybe you meant ProtocolId?

Reply:
Oooh, I'm just trying to imagine how one might debug an RPC, one approach is to println the RPC through the system and to do that you need to print the distinct parts and I suspect in our case that's the peer and request id?